persistent and ephemeral csi volumes

pohly opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This is part of the effort to move CSI inline ephemeral volumes to beta.

**Which issue(s) this PR fixes**:
Related-to: https://github.com/kubernetes/kubernetes/issues/79624

**Does this PR introduce a user-facing change?**:
```release-note
when PodInfoOnMount is enabled for a CSI driver, the new csi.storage.k8s.io/ephemeral parameter in the volume context allows a driver's NodePublishVolume implementation to determine on a case-by-case basis whether the volume is ephemeral or a normal persistent volume
```

pohly committed:
CSI: allow drivers that can handle persistent and ephemeral volumes

The conceptual change is that the mode in which a volume gets handled
is derived from it's spec, not from the ability of the driver. In
practice, that is already how the code worked because it didn't
actually look at CSIDriver.Spec.Mode at all.

Therefore the code change itself is mostly just renaming "driver mode"
to "volume mode". In some places (CanDeviceMount, CanAttach) the
feature check that was used elsewhere seemed to be missing. Now their
code path for ephemeral volumes are also only entered if that feature
is enabled.

The sanity check whether a CSI driver is being used correctly still
needs to be implemented.

Related-to: https://github.com/kubernetes/kubernetes/issues/79624

pohly commented:
/assign @vladimirvivien 
/sig-storage

pohly commented:
/sig storage
/priority important

cwdsuzhou commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -60,7 +61,7 @@ type csiMountMgr struct {
 	plugin         *csiPlugin
 	driverName     csiDriverName
-	driverMode     driverMode
+	volumeMode     volumeMode
```
If

pohly replied:
Yes, naming is hard :frowning:

I also considered `ephemeralMode` or `inlineMode`, but both then don't make much sense when the value is `persistent`.

Any suggestions?

pohly replied:
How about `VolumeSource`, with `persistent` and `ephemeral` as values?

It's then different from `driverMode`, but perhaps that's okay.

kfox1111 replied:
Yeah, naming things is hard. Slight preference to the previous name, driverMode, as that kind of matches the spec:
CSIDriver.Spec.Mode -> driverMode.

Or maybe csiDriverMode?

pohly replied:
`VolumeSource` also could be misunderstood (see snapshotting...). Let's see how well `CSIVolumeMode` works...

kfox1111 replied:
Works for me.

pohly replied:
I left "driverMode" unchanged, because as you said, it is derived from `CSIDriver.Spec.Mode`.

pohly committed:
updated Bazel files

kfox1111 reviewed (commented):
With my limited knowledge of how e2e works, and excluding the naming question, this all looks good to me. :)

cwdsuzhou commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -603,14 +614,17 @@ func (p *csiPlugin) CanAttach(spec *volume.Spec) (bool, error) {
+		if csiVolumeMode == ephemeralVolumeMode {
+			klog.V(5).Info(log("plugin.CanDeviceMount skipped ephemeral mode detected for spec %v", spec.Name()))
+			return false, nil
+		}
```
Would this make it more clear?

If you do not think so, ignore it.

```
if !inlineEnabled {
    return true, nil
}
driverMode, err := p.getDriverMode(spec)
...

return true, nil

```

pohly replied:
Yes, I think that's better and more consistent with usual Go style. I changed it, please have another look.

pohly commented:
/retest

pohly commented:
@cwdsuzhou is it okay now?

pohly commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -323,6 +324,7 @@ func (c *csiMountMgr) podAttributes() (map[string]string, error) {
 		"csi.storage.k8s.io/pod.namespace":       c.pod.Namespace,
 		"csi.storage.k8s.io/pod.uid":             string(c.pod.UID),
 		"csi.storage.k8s.io/serviceAccount.name": c.pod.Spec.ServiceAccountName,
+		"csi.storage.k8s.io/ephemeral":           strconv.FormatBool(c.csiVolumeMode == ephemeralVolumeMode),
```
@kfox1111 should the presence of this new field in the PodInfo depend on the feature gate?

kfox1111 replied:
Hmm... That is a good question...

It shouldn't hurt anything to have it always defined as it is, but following the principal that nothing should change if the gate is disabled, it probably should be behind the gate, yeah.

pohly replied:
I've made it optional, i.e. it only gets added when the feature gate is set.

pohly commented on test/e2e/storage/testsuites/ephemeral.go:
```diff
@@ -0,0 +1,230 @@
+func InitEphemeralTestSuite() TestSuite {
+	return &ephemeralTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "ephemeral",
```
/hold

Currently ephemeral inline volumes are an alpha feature, so this whole suite must have a `[Feature:CSIInlineVolume]` tag. Will add it.

pohly replied:
Done.

cwdsuzhou commented:
> @cwdsuzhou is it okay now?

I think it is okay now.

LGTM

pohly committed:
e2e storage: csi-mock tests for ephemeral inline volumes

The PodInfo tests can be extended to also cover the new
csi.storage.k8s.io/ephemeral flag. However, the presence of that flag
depends on whether inline volume support is enabled, so tests that run
with and without the feature have to detect that at runtime.

Other tests have a feature tag and thus can assume that they only run
when that feature is enabled. However, we need a newer csi-mock driver
before we can actually ask it to publish an ephemeral inline volume.

pohly commented:
In order to test the new csi.storage.k8s.io/ephemeral properly, I ended up extending the csi_mock_volume tests. However, now this PR depends on a pending PRs against csi-test (https://github.com/kubernetes-csi/csi-test/pull/209) and a new release.

/hold

vladimirvivien commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -77,6 +77,12 @@ type driverMode string
 
 const persistentDriverMode driverMode = "persistent"
 const ephemeralDriverMode driverMode = "ephemeral"
+const combinedDriverMode driverMode = "persistent+ephemeral"
```
I have missed most recent discussions, is this a new mode being introduced where driver can operate in ether mode ?

pohly replied:
Yes, exactly. KEP update is here: https://github.com/kubernetes/enhancements/pull/1129/files

Hostpath already implements that and supports both modes now: https://github.com/kubernetes-csi/csi-driver-host-path/pull/67

msau42 replied:
These consts should be defined in the api. If the API is not there yet, then we should remember to remove these once we add it

vladimirvivien reviewed (commented):
LGTM
Great effort on the e2e

vladimirvivien commented:
/lgtm

pohly commented:
@vladimirvivien I had to rebase and fix a unit test failure that came from making the "ephemeral" podinfo field also depend on the feature flag. Please re-add the LGTM.

pohly commented:
/cancel hold

mock-driver v2.1.0 exists now, so this can be merged.

/test gce-alpha-features

vladimirvivien commented:
/lgtm

msau42 commented on test/e2e/storage/testpatterns/testpattern.go:
```diff
@@ -44,6 +44,8 @@ var (
 	// DynamicPV represents a volume type for dynamic provisioned Persistent Volume
 	DynamicPV TestVolType = "DynamicPV"
+	// CSIVolume represents a volume type that is defined inline and provided by a CSI driver.
+	CSIVolume TestVolType = "CSIVolume"
```
Prefer having inline in the name so that it's not confused with pvs.

Is the reason why this needs to be a new field is because current inline volume pattern assumes persistence?

pohly replied:
I can rename "CSIVolume" to "CSIInlineVolume". I chose "CSIVolume" because that's also the term used in "CSIVolumeSource" (https://github.com/kubernetes/kubernetes/blob/1dac5fd14a54ac4972339dbe55f9f03688fd7542/staging/src/k8s.io/api/core/v1/types.go#L1706-L1707), but I agree, the longer term is better and also used elsewhere (for example, in the feature gate).

The reason for the separate volume type is, as you said, that the characteristics of the volume are different. I don't know whether it makes a difference in practice, though.

msau42 replied:
Actually is this used anywhere?  If not, then let's leave it out for now.

The test pattern here is used to run the same test cases against different configurations.  But we have explicit test cases for csi inline, so this is not needed for now.

msau42 replied:
Nevermind I see it now. I was looking at the wrong commit.

pohly replied:
I checked: it is needed in base.go to determine whether the driver supports the test pattern. See https://github.com/kubernetes/kubernetes/blob/aa8999a37dda4c5e430fb35c80a45d2c19ee4ad8/test/e2e/storage/testsuites/base.go#L137-L138

pohly replied:
> But we have explicit test cases for csi inline, so this is not needed for now.

At least one test pattern is needed, otherwise the storage testsuite framework will not instantiate tests.

pohly replied:
The name change was pushed.

pohly commented:
/hold cancel

pohly reviewed (commented):
> These consts should be defined in the api. If the API is not there yet, then we should remember to remove these once we add it

The API still needs to be added. I've done that in a follow-up branch and also moved the constants.

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
I included another fix (problem found while enabling in-tree testing of ephemeral volumes with the CSI hostpath driver):

```
diff --git a/test/e2e/storage/testsuites/ephemeral.go b/test/e2e/storage/testsuites/ephemeral.go
index bc7b8b244e..c9e4152cae 100644
--- a/test/e2e/storage/testsuites/ephemeral.go
+++ b/test/e2e/storage/testsuites/ephemeral.go
@@ -92,7 +92,7 @@ func (p *ephemeralTestSuite) defineTests(driver TestDriver, pattern testpatterns
                l.testCase = &EphemeralTest{
                        Client:     l.config.Framework.ClientSet,
                        Namespace:  f.Namespace.Name,
-                       DriverName: dInfo.Name,
+                       DriverName: l.config.GetUniqueDriverName(),
                        Node:       framework.NodeSelection{Name: l.config.ClientNodeName},
                        GetVolumeAttributes: func(volumeNumber int) map[string]string {
                                return eDriver.GetVolumeAttributes(l.config, volumeNumber)
```

It worked when testing an external driver because the unique name is the same as the base name...

pohly commented:
> It worked when testing an external driver because the unique name is the same as the base name...

And using the generated name breaks testing an external driver... this needs some more work before it works for both.

/hold

pohly committed:
e2e: remove unused ShortName from external driver definition

The name was meant to be used as shorter replacement for potentially
long CSI driver names, but was never used in practice.

pohly commented:
Fixed this way:
```
diff --git a/test/e2e/storage/external/external.go b/test/e2e/storage/external/external.go
index 6ff28829de..cb2c95535b 100644
--- a/test/e2e/storage/external/external.go
+++ b/test/e2e/storage/external/external.go
@@ -148,9 +148,6 @@ type driverDefinition struct {
 	// the default file system are enabled.
 	DriverInfo testsuites.DriverInfo
 
-	// ShortName is used to create unique names for test cases and test resources.
-	ShortName string
-
 	// StorageClass must be set to enable dynamic provisioning tests.
 	// The default is to not run those tests.
 	StorageClass struct {
@@ -186,6 +183,8 @@ type driverDefinition struct {
 	// has to be defined to enable testing of inline ephemeral volumes.
 	// If a test needs more volumes than defined, some of the defined
 	// volumes will be used multiple times.
+	//
+	// DriverInfo.Name is used as name of the driver in the inline volume.
 	InlineVolumeAttributes []map[string]string
 
 	// ClaimSize defines the desired size of dynamically
@@ -301,6 +300,10 @@ func (d *driverDefinition) GetVolumeAttributes(config *testsuites.PerTestConfig,
 	return d.InlineVolumeAttributes[volumeNumber%len(d.InlineVolumeAttributes)]
 }
 
+func (d *driverDefinition) GetCSIDriverName(config *testsuites.PerTestConfig) string {
+	return d.DriverInfo.Name
+}
+
 func (d *driverDefinition) PrepareTest(f *framework.Framework) (*testsuites.PerTestConfig, func()) {
 	config := &testsuites.PerTestConfig{
 		Driver:         d,
diff --git a/test/e2e/storage/testsuites/ephemeral.go b/test/e2e/storage/testsuites/ephemeral.go
index c9e4152cae..ee9cfcbf71 100644
--- a/test/e2e/storage/testsuites/ephemeral.go
+++ b/test/e2e/storage/testsuites/ephemeral.go
@@ -92,7 +92,7 @@ func (p *ephemeralTestSuite) defineTests(driver TestDriver, pattern testpatterns
 		l.testCase = &EphemeralTest{
 			Client:     l.config.Framework.ClientSet,
 			Namespace:  f.Namespace.Name,
-			DriverName: l.config.GetUniqueDriverName(),
+			DriverName: eDriver.GetCSIDriverName(l.config),
 			Node:       framework.NodeSelection{Name: l.config.ClientNodeName},
 			GetVolumeAttributes: func(volumeNumber int) map[string]string {
 				return eDriver.GetVolumeAttributes(l.config, volumeNumber)
diff --git a/test/e2e/storage/testsuites/testdriver.go b/test/e2e/storage/testsuites/testdriver.go
index 74224ca8d4..4c05d2554f 100644
--- a/test/e2e/storage/testsuites/testdriver.go
+++ b/test/e2e/storage/testsuites/testdriver.go
@@ -111,6 +111,11 @@ type EphemeralTestDriver interface {
 	// all be the same or different, depending what the driver supports
 	// and/or wants to test.
 	GetVolumeAttributes(config *PerTestConfig, volumeNumber int) map[string]string
+
+	// GetCSIDriverName returns the name that was used when registering with
+	// kubelet. Depending on how the driver was deployed, this can be different
+	// from DriverInfo.Name.
+	GetCSIDriverName(config *PerTestConfig) string
 }
 
 // SnapshottableTestDriver represents an interface for a TestDriver that supports DynamicSnapshot
```

The removal of `ShortName` is unrelated and a separate commit, I just noticed while looking into the driver naming problem that it exists although not used anywhere.

pohly commented:
/hold cancel

pohly commented:
For reference, the in-tree usage of the new ephemeral testsuite test is part of PR #80568, currently in commit 4de69dcf9e759c9f096f83d0a444a3e091342f73.

pohly commented:
/test pull-kubernetes-integration

msau42 commented:
/lgtm
/approve
